File: app/src/main/java/cat/xojan/random1/domain/interactor/PodcastDataInteractor.java
package cat.xojan.random1.domain.interactor;

import android.content.Context;
import android.os.Environment;
import android.util.Log;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import cat.xojan.random1.domain.model.Podcast;
import cat.xojan.random1.domain.model.Program;
import cat.xojan.random1.domain.model.Section;
import cat.xojan.random1.domain.repository.PodcastRepository;
import rx.Observable;
import rx.Subscriber;
import rx.subjects.PublishSubject;

public class PodcastDataInteractor {

    private static final int NUM_PODCASTS = 30;
    private static final String TAG = PodcastDataInteractor.class.getSimpleName();
    public static final String SEPARATOR = "_";
    public static final String EXTENSION = ".mp3";
    public static final String IMAGE = "programtitle";

    private final PodcastRepository mPodcastRepo;
    private final Context mContext;
    private PublishSubject<List<Podcast>> mDownloadedPodcastsSubject = PublishSubject.create();
    private List<Podcast> mLatestPodcasts;
    private List<Podcast> mPodcastsByProgram;
    private List<Podcast> mPodcastsBySection;
    private Program mProgram;
    private Section mSection;

    @Inject
    public PodcastDataInteractor(PodcastRepository podcastRepository, Context context) {
        mPodcastRepo = podcastRepository;
        mContext = context;
    }

    public Observable<List<Podcast>> loadLatestPodcasts(final boolean refresh) {
        return Observable.create(new Observable.OnSubscribe<List<Podcast>>() {
            @Override
            public void call(Subscriber<? super List<Podcast>> subscriber) {
                try {
                    if (mLatestPodcasts == null || refresh) {
                        mLatestPodcasts = mPodcastRepo.getLatestPodcasts(NUM_PODCASTS);
                    }
                    subscriber.onNext(mLatestPodcasts);
                    subscriber.onCompleted();
                } catch (IOException e) {
                    subscriber.onError(e);
                }
            }
        });
    }

    public Observable<List<Podcast>> loadPodcastsByProgram(final Program program,
                                                           final boolean refresh) {
        return Observable.create(new Observable.OnSubscribe<List<Podcast>>() {
            @Override
            public void call(Subscriber<? super List<Podcast>> subscriber) {
                try {
                    if (mPodcastsByProgram == null || refresh || !mProgram.equals(program)) {
                        mProgram = program;
                        mPodcastsByProgram = mPodcastRepo.getLatestPodcasts(NUM_PODCASTS,
                                program.getParam());
                    }
                    subscriber.onNext(mPodcastsByProgram);
                    subscriber.onCompleted();
                } catch (IOException e) {
                    subscriber.onError(e);
                }
            }
        });
    }

    public Observable<List<Podcast>> loadPodcastsBySection(final Section section,
                                                        final boolean refresh) {
        return Observable.create(new Observable.OnSubscribe<List<Podcast>>() {
            @Override
            public void call(Subscriber<? super List<Podcast>> subscriber) {
                try {
                    if (mPodcastsBySection == null || refresh || !mSection.equals(section)) {
                        mSection = section;
                        mPodcastsBySection = mPodcastRepo.getLatestSections(NUM_PODCASTS,
                                section.getParam());
                    }
                    subscriber.onNext(mPodcastsBySection);
                    subscriber.onCompleted();
                } catch (IOException e) {
                    subscriber.onError(e);
                }
            }
        });
    }

    public void refreshDownloadedPodcasts() {
        try {
            List<Podcast> podcastList = new ArrayList<>();
            getDownloading(podcastList);
            int downloading = podcastList.size();
            Log.d(TAG, "downloading: " + downloading);
            getDownloaded(podcastList);
            int downloaded = podcastList.size() - downloading;
            Log.d(TAG, "downloaded: " + downloaded);

            mDownloadedPodcastsSubject.onNext(podcastList);
        } catch (Exception e) {
            mDownloadedPodcastsSubject.onError(e);
        };
    }

    public void addDownload(String category, String description, String programTitle) {
        File from = new File(mContext.getExternalFilesDir(Environment.DIRECTORY_DOWNLOADS) +
                File.separator + category + SEPARATOR + description + EXTENSION
                + IMAGE + programTitle);
        File to = new File(mContext.getExternalFilesDir(Environment.DIRECTORY_PODCASTS) +
                File.separator + category + SEPARATOR + description + EXTENSION
                + IMAGE + programTitle);

        if (from.renameTo(to)) {
            Log.d(TAG, "moving download from " + from.getPath() + " to " + to.getPath());
        } else {
            from.delete();
        }
        refreshDownloadedPodcasts();
    }

    public void deleteDownload(Podcast podcast) {
        File file = new File(podcast.getFilePath());
        if (file.delete()) {
            refreshDownloadedPodcasts();
        }
    }

    public Observable<List<Podcast>> getDownloadedPodcasts() {
        return mDownloadedPodcastsSubject;
    }

    private void getDownloaded(List<Podcast> podcastList) {
        File iternalFileDirectory = mContext.getExternalFilesDir(Environment.DIRECTORY_PODCASTS);
        for (File podcastFile : iternalFileDirectory.listFiles()) {
            addPodcastToList(podcastList, podcastFile, Environment.DIRECTORY_PODCASTS,
                    Podcast.State.DOWNLOADED);
        }
    }

    private void getDownloading(List<Podcast> podcastList) {
        File iternalFileDirectory = mContext.getExternalFilesDir(Environment.DIRECTORY_DOWNLOADS);
        for (File podcastFile : iternalFileDirectory.listFiles()) {
            addPodcastToList(podcastList, podcastFile, Environment.DIRECTORY_DOWNLOADS,
                    Podcast.State.DOWNLOADING);
        }
    }

    private void addPodcastToList(List<Podcast> podcastList, File podcastFile, String directory,
                                  Podcast.State podcastState) {
        String[] fileName = podcastFile.getPath()
                .split(directory + "/")[1].split(SEPARATOR);
        String category = fileName[0];
        String description = fileName[1].split(IMAGE)[0].replace(EXTENSION, "");

        String programTitle = null;
        if (fileName[1].split(IMAGE).length > 1) {
            programTitle = fileName[1].split(IMAGE)[1];
        }

        Podcast podcast = new Podcast(category, description, podcastFile.getPath(),
                podcastState, programTitle);
        podcastList.add(podcast);
    }
}

File: app/src/main/java/cat/xojan/random1/domain/interactor/PodcastDataInteractor.kt
package cat.xojan.random1.domain.interactor

import android.app.DownloadManager
import android.content.Context
import android.os.Environment
import android.support.v4.media.MediaBrowserCompat
import android.support.v4.media.MediaDescriptionCompat
import android.text.TextUtils
import android.util.Log
import cat.xojan.random1.domain.model.EventLogger
import cat.xojan.random1.domain.model.Podcast
import cat.xojan.random1.domain.model.Podcast.Companion.PODCAST_DATE
import cat.xojan.random1.domain.model.Podcast.Companion.PODCAST_DOWNLOAD_REFERENCE
import cat.xojan.random1.domain.model.Podcast.Companion.PODCAST_FILE_PATH
import cat.xojan.random1.domain.model.Podcast.Companion.PODCAST_STATE
import cat.xojan.random1.domain.model.PodcastState
import cat.xojan.random1.domain.model.Program
import cat.xojan.random1.domain.repository.DownloadPodcastRepository
import cat.xojan.random1.domain.repository.PodcastPreferencesRepository
import cat.xojan.random1.domain.repository.PodcastRepository
import cat.xojan.random1.domain.repository.ProgramRepository
import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.functions.BiFunction
import io.reactivex.subjects.PublishSubject
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.util.*
import javax.inject.Inject

class PodcastDataInteractor @Inject constructor(
        private val programRepo: ProgramRepository,
        private val podcastRepo: PodcastRepository,
        private val podcastPref: PodcastPreferencesRepository,
        private val downloadManager: DownloadManager,
        private val context: Context,
        private val downloadRepo: DownloadPodcastRepository,
        private val eventLogger: EventLogger)
{

    companion object {
        const val EXTENSION = ".mp3"
    }
    private val TAG = PodcastDataInteractor::class.simpleName

    private val stateUpdatesSubject: PublishSubject<List<MediaBrowserCompat.MediaItem>> =
            PublishSubject.create()

    fun getHourByHourPodcasts(programId: String, refresh: Boolean): Single<List<Podcast>> {
        val program = programRepo.getProgram(programId)
        val podcastList = podcastRepo.getPodcasts(programId, null, refresh)

        return Single.zip(program, podcastList,
                BiFunction<Program, List<Podcast>, List<Podcast>> { pr, podList ->
            for (p in podList) {
                p.programId = pr.id
                p.smallImageUrl = pr.smallImageUrl
                p.bigImageUrl = pr.bigImageUrl
            }
                    podList
        })
    }

    fun getSectionPodcasts(programId: String, sectionId: String, refresh: Boolean):
            Single<List<Podcast>> {
        val program = programRepo.getProgram(programId)
        val podcastList = podcastRepo.getPodcasts(programId, sectionId, refresh)

        return Single.zip(program, podcastList,
                BiFunction<Program, List<Podcast>, List<Podcast>> { pr, podList ->
                    for (p in podList) {
                        p.programId = pr.id
                        p.smallImageUrl = pr.smallImageUrl
                        p.bigImageUrl = pr.bigImageUrl
                    }
                    podList
                })
    }

    fun isSectionSelected(): Boolean {
        return podcastPref.isSectionSelected()
    }

    fun setSectionSelected(selected: Boolean) {
        podcastPref.setSectionSelected(selected)
    }

    fun download(podcast: MediaDescriptionCompat) {
        val uri = podcast.mediaUri
        val request = DownloadManager.Request(uri)
                .setTitle(podcast.title)
                .setDestinationInExternalFilesDir(
                        context,
                        Environment.DIRECTORY_DOWNLOADS,
                        podcast.mediaId + EXTENSION)
                .setVisibleInDownloadsUi(true)

        val reference = downloadManager.enqueue(request)
        podcast.extras?.putLong(PODCAST_DOWNLOAD_REFERENCE, reference)
        downloadRepo.addDownloadingPodcast(podcast)
    }

    fun deleteDownload(podcast: MediaDescriptionCompat) {
        val file = File(podcast.extras?.getString(PODCAST_FILE_PATH))
        if (file.delete()) {
            downloadRepo.deleteDownloadedPodcast(podcast)
        }
    }

    fun getDownloadedPodcasts(): Single<List<MediaBrowserCompat.MediaItem>> {
        return Single.just(fetchDownloadedPodcasts())
                .flatMap {
                    podcasts -> Observable.just(podcasts)
                        .flatMapIterable { p -> p }
                        .filter { p -> p.description.extras?.getSerializable(PODCAST_STATE) ==
                                PodcastState.DOWNLOADED }
                        .sorted { p1, p2 -> (p2.description.extras?.getSerializable(PODCAST_DATE)
                                as Date).compareTo(p1.description.extras?.getSerializable
                            (PODCAST_DATE) as Date) }
                        .toList()
                }
    }

    fun getPodcastStateUpdates(): PublishSubject<List<MediaBrowserCompat.MediaItem>> {
        return stateUpdatesSubject
    }

    fun refreshDownloadedPodcasts() {
        stateUpdatesSubject.onNext(fetchDownloadedPodcasts())
    }

    fun fetchDownloadedPodcasts(): List<MediaBrowserCompat.MediaItem> {
        val podcastList = HashSet<MediaDescriptionCompat>()
        val downloading = downloadRepo.getDownloadingPodcasts()
        val downloaded = downloadRepo.getDownloadedPodcasts()
        podcastList.addAll(downloading)
        podcastList.addAll(downloaded)

        Log.d(TAG, "Downloading: " + downloading.size + ", downloaded: " + downloaded.size)

        return podcastList.map { description ->
            MediaBrowserCompat.MediaItem(description, MediaBrowserCompat.MediaItem.FLAG_PLAYABLE) }
    }

    fun deleteDownloading(reference: Long) {
        val podcast: MediaDescriptionCompat? = downloadRepo.getDownloadingPodcasts()
                .lastOrNull {
                    reference == it.extras?.getLong(PODCAST_DOWNLOAD_REFERENCE)
                }
        podcast?.let {
            downloadRepo.deleteDownloadingPodcast(podcast)
        }
    }

    fun addDownload(audioId: String) {
        val from = File(context.getExternalFilesDir(Environment.DIRECTORY_DOWNLOADS).toString() +
                File.separator + audioId + EXTENSION)
        val to = File(context.getExternalFilesDir(Environment.DIRECTORY_PODCASTS).toString() +
                File.separator + audioId + EXTENSION)

        if (from.renameTo(to)) {
            Log.d(TAG, "moving download from " + from.path + " to " + to.path)
            downloadRepo.setPodcastAsDownloaded(audioId, to.path)
        } else {
            from.delete()
            to.delete()
        }
    }

    fun exportPodcasts(): Single<Unit> {
        eventLogger.logExportPodcastsAction()
        return Single.create { subscriber ->
            try {
                val iternalFileDir = context.getExternalFilesDir(Environment.DIRECTORY_PODCASTS)
                val externalFilesDir = Environment.getExternalStoragePublicDirectory(
                        Environment.DIRECTORY_PODCASTS)

                externalFilesDir.mkdirs()

                for (podcastFile in iternalFileDir!!.listFiles()) {
                    val audioId = podcastFile.path
                            .split((Environment.DIRECTORY_PODCASTS + "/").toRegex())
                            .dropLastWhile({ it.isEmpty() })
                            .toTypedArray()[1].replace(".mp3", "")
                    var podcastTitle = getDownloadedPodcastTitle(audioId)

                    if (!TextUtils.isEmpty(podcastTitle)) {
                        podcastTitle = podcastTitle!!.replace("/", "-")
                        val dest = File(externalFilesDir, "$podcastTitle.mp3")
                        copy(podcastFile, dest)
                    }
                    eventLogger.logExportedPodcast(audioId, podcastTitle)
                }
                subscriber.onSuccess(Unit)
                eventLogger.logExportedPodcastsSuccess()
            } catch (e: Throwable) {
                subscriber.onError(e)
                eventLogger.logExportedPodcastsFail()
            }
        }
    }

    private fun getDownloadedPodcastTitle(audioId: String): String? {
        return downloadRepo.getDownloadedPodcastTitle(audioId)
    }

    private fun copy(src: File, dst: File) {
        val inStream = FileInputStream(src)
        val outStream = FileOutputStream(dst)
        val inChannel = inStream.channel
        val outChannel = outStream.channel
        inChannel.transferTo(0, inChannel.size(), outChannel)
        inStream.close()
        outStream.close()
    }
}